src/models/model.cpp

```diff
@@ -271,167 +281,193 @@ void Model::InitDeviceAllocator([[maybe_unused]] OrtSession& session) {
+    }
 
-      std::vector<const char*> keys, values;
-      for (auto& option : provider_options.options) {
-        keys.emplace_back(option.first.c_str());
-        values.emplace_back(option.second.c_str());
-      }
+    for (auto& provider_options : config_session_options.provider_options) {
+      if (provider_options.name == "cuda") {
+        auto ort_provider_options = OrtCUDAProviderOptionsV2::Create();
+        std::vector<const char*> keys, values;
+        for (auto& option : provider_options.options) {
+          keys.emplace_back(option.first.c_str());
+          values.emplace_back(option.second.c_str());
+        }
+        ort_provider_options->Update(keys.data(), values.data(), keys.size());
+
+        // Create and set our cudaStream_t
+        if (!cuda_stream_.get())
+          cuda_stream_.Create();
+
+        ort_provider_options->UpdateValue("user_compute_stream", cuda_stream_.get());
+
+        session_options.AppendExecutionProvider_CUDA_V2(*ort_provider_options);
+        // Device type determines the scoring device.
+        // Only use the primary session options to determine the device type
+        if (is_primary_session_options)
+          device_type_ = DeviceType::CUDA;  // Scoring will use CUDA
+      } else if (provider_options.name == "rocm") {
+        OrtROCMProviderOptions ort_provider_options;
+
+        std::vector<const char*> keys, values;
+        for (auto& option : provider_options.options) {
+          keys.emplace_back(option.first.c_str());
+          values.emplace_back(option.second.c_str());
+        }
 
-      Ort::ThrowOnError(Ort::api->UpdateROCMProviderOptions(&ort_provider_options, keys.data(), values.data(), keys.size()));
-      ort_options.AppendExecutionProvider_ROCM(ort_provider_options);
+        Ort::ThrowOnError(Ort::api->UpdateROCMProviderOptions(&ort_provider_options, keys.data(), values.data(), keys.size()));
+        session_options.AppendExecutionProvider_ROCM(ort_provider_options);
 #if USE_DML
-    } else if (provider_options.name == "dml") {
-      auto current_module_path = CurrentModulePath();
-      dml_objects_ = DmlHelpers::CreateDmlObjects(current_module_path);
-
-      constexpr auto directml_dll = "DirectML.dll";
-      wil::unique_hmodule smart_directml_dll(LoadLibraryEx(directml_dll, nullptr, 0));
-      THROW_LAST_ERROR_IF(!smart_directml_dll);
-
-      if (LoadLibraryEx(directml_dll, nullptr, 0) == NULL) {
-        throw std::runtime_error("DirectML.dll not found");
-      }
-
-      auto dml_create_device1_fn = reinterpret_cast<decltype(&DMLCreateDevice1)>(GetProcAddress(smart_directml_dll.get(), "DMLCreateDevice1"));
-      THROW_LAST_ERROR_IF(!dml_create_device1_fn);
-      THROW_IF_FAILED(dml_create_device1_fn(dml_objects_.d3d12_device.Get(), DML_CREATE_DEVICE_FLAG_NONE, DML_FEATURE_LEVEL_5_0, IID_PPV_ARGS(&dml_device_)));
+      } else if (provider_options.name == "dml") {
+        if (!p_dml_api_) {
```

Comment:
>if (!p_dml_api_) { [](http://example.com/codeflow?start=8&length=18)

why add this check? the p_dmp_api_ is set below.

Reply:
All the pipeline model session options are created by calling this lambda function multiple times (once for each pipeline session options).
However, this lambda function not only updates the session options passed in, it also sets members on `this` `Model`. Particularly, it sets the `device_type_` and the `p_dml_api_` on this. These members once created should not be updated again. So, I check if the members are already created. If not, we create them. If they are, we skip this block of code.